daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala

```diff
@@ -222,7 +222,7 @@ class Engine(val config: EngineConfig = new EngineConfig(LanguageVersion.StableV
       (rtx, _) = result
       validationResult <-
         transaction.Validation
-          .isReplayedBy(tx, rtx)
+          .isReplayedBy(transaction.Normalization.normalizeTx(tx), rtx)
```

Comment:
Why do we normalize here? Shouldn’t the submitted transaction already be serialized?

Reply:
I'm not sure we can rely on this. If we remove this call to `normalizeTx` then we get test failures.
For example:
```
bazel run //ledger/ledger-on-memory:ledger-on-memory-tests_test_suite_src_test_suite_scala_com_daml_ledger_on_memory_InMemoryLedgerReaderWriterIntegrationSpec.scala`
```

Tracing the stack shows the call to `Engine.validate` comes from `ModelConformanceValidator` from `TransactionCommitter`, and ultimately from `PreExecutingSubmissionValidator`.

I assume the TX has not been serialized at this point.

Reply:
Hm interesting, do you have a list of test failures? Wondering if it’s reasonable to insert normalization there in particular if it’s things like ledger-on-memory where we don’t care about performance rather than unnecessarily normalizing in places where performance might matter.

Reply:
I just pushed a commit to remove the call to `normalizeTx` from `Engine.validate`, so CI can give us the full list...

All of the failures I have seen locally (so far) have looked like this:
```
The partial function passed as the second parameter to inside was not defined at the value passed as the first parameter to inside, which was: CommandRejected(2021-08-16T10:04:37.215Z,SubmitterInfo(List(alice),tests,X1,2021-08-16T10:04:47.158Z),Disputed(recreated and original transaction mismatch VersionedTransaction(V11,HashMap(),ImmArray()) expected, but VersionedTransaction(V10,Map(),ImmArray()) is recreated)) (ParticipantStateIntegrationSpecBase.scala:780)
```
So the mismatch is:
`VersionedTransaction(V11,HashMap(),ImmArray())`
vs
`VersionedTransaction(V10,Map(),ImmArray())`

This is quite interesting because the correct version for an empty TX is `V10`. But somewhere in the code is creating them at `V11`! Just need to find where this is happening...

Reply:
The above badly version empty TX issue causes fails in the following 4 tests:
```
FAIL: //ledger/ledger-on-sql:ledger-on-sql-tests_test_suite_src_test_suite_scala_com_daml_ledger_on_sql_SqliteMemorySqlLedgerReaderWriterIntegrationSpec.scala
FAIL: //ledger/ledger-on-sql:ledger-on-sql-tests_test_suite_src_test_suite_scala_com_daml_ledger_on_sql_PostgresqlSqlLedgerReaderWriterIntegrationSpec.scala
FAIL: //ledger/ledger-on-sql:ledger-on-sql-tests_test_suite_src_test_suite_scala_com_daml_ledger_on_sql_H2MemorySqlLedgerReaderWriterIntegrationSpec.scala
FAIL: //ledger/ledger-on-memory:ledger-on-memory-tests_test_suite_src_test_suite_scala_com_daml_ledger_on_memory_InMemoryLedgerReaderWriterIntegrationSpec.scala
```

Reply:
CI (windows/macos) confirms it is just these 4 failures.

Reply:
Interesting, wouldn’t have expected ledger on sql to fail. Fine to normalize for now and then drop it once we normalize within the engine.

Reply:
:+1: